### DISCUSSION TITLE
Blazor OnPropertyChanged and Custom AddBusinessRules

### DISCUSSION THREAD

**Discussion by jhuerta3** [MarimerLLC/csla#3230] [Questions] [answered]
Hello all,

Im have been trying to figure this out. I cant seem to call the database from the BusinessEdit object OnPropertyChange method and also from AddBusinessRules method. I keep getting the following error, "Synchronous data access is not supported on this runtime. Please use async data access methods"  I tried several things but no luck. I tried to change the OnPropertyChange to async but I could not find any methods to convert them to async. This works fine from the razor pages since they are all async. So basically the question is how do I access the database from the server side/business objects. Im not sure if the ApplicationContext.GetRequiredService is the correct way to do it. 

I will keep doing some research. Thank you. 

        protected override void OnPropertyChanged(IPropertyInfo propertyInfo)
        {
            base.OnPropertyChanged(propertyInfo);
            if (propertyInfo == StatusIDProperty)
            {
                if (StatusID.ToString() != null && StatusID.ToString() != "00000000-0000-0000-0000-000000000000")
                {
                    LookUpTableInfoPortal = ApplicationContext.GetRequiredService<IDataPortal<LookUpTableInfo>>();
                    LookUpTableFactory objLookFactory = new LookUpTableFactory(LookUpTableInfoPortal);
                    LookUpTableInfo objStatus = objLookFactory.GetkLookUpInfo(StatusID);

                    if (objStatus.LookUpName.Equals(SystemEnums.EquipmentStatusEnum.Obsolete) || 
                        objStatus.LookUpName.Equals(SystemEnums.EquipmentStatusEnum.Damaged))
                        ActiveFlag = false;
                    else
                        ActiveFlag = true;
                }
            }
        }

and

        protected override void AddBusinessRules()
        {
            LookUpTableInfoPortal = ApplicationContext.GetRequiredService<IDataPortal<LookUpTableInfo>>();
            base.AddBusinessRules();
            BusinessRules.AddRule(new CheckStatusChange(StatusIDProperty, ActiveFlag, LookUpTableInfoPortal));
        }

and 

    public class CheckStatusChange : BusinessRule
    {
        Csla.IDataPortal<AssetMgmtSystem.Shared.LookUpTableInfo> LookUpTableInfoPortal;
        public bool ActiveFlagStatus { get; set; }

        public CheckStatusChange(Csla.Core.IPropertyInfo primaryProperty, 
                                 bool activeflagstatus, 
                                 Csla.IDataPortal<App.Shared.LookUpTableInfo> lookuttableinfoportal)
          : base(primaryProperty)
        {
            ActiveFlagStatus = activeflagstatus;
            LookUpTableInfoPortal = lookuttableinfoportal;
        }

        protected override void Execute(IRuleContext context)
        {
            var StatusIDValue = (Guid)ReadProperty(context.Target, PrimaryProperty);

                if (StatusIDValue.ToString() != null && StatusIDValue.ToString() != "00000000-0000-0000-0000-000000000000")
                {
                    LookUpTableFactory objLookFactory = new LookUpTableFactory(LookUpTableInfoPortal);
                    LookUpTableInfo objStatus = objLookFactory.GetkLookUpInfo(StatusIDValue);

                    if ((objStatus.LookUpName.Equals(SystemEnums.EquipmentStatusEnum.Obsolete) || objStatus.LookUpName.Equals(SystemEnums.EquipmentStatusEnum.Damaged)) && ActiveFlagStatus == true)
                        context.AddWarningResult("Check avilability. Active flag should be false for status.");
                }
            return;
        }
    }

**Comment by Chicagoan2016** [comment]
I didn't understand your 'use case' : ) but here is something I have done in the past (no DI back then)
In my business rule class (inside Execute method) I called either a command object or readonly collection object to get values I needed.

**Comment by rockfordlhotka** [comment]
The exception is telling you that your code is running in a .NET runtime environment where synchronous server calls aren't supported, so your data portal calls need to be async. It isn't about your _calling_ code (what you are showing here), it is about the _called_ code - specifically where you are actually using the data portal.

Your `GetkLookUpInfo` method must be doing a `Fetch` instead of `FetchAsync` - or something like that.

**Reply by jhuerta3** [reply]
Hello Both, I see what you mean. Yes my GetLookUpInfo was using the following 

        public LookUpTableInfo GetkLookUpInfo(Guid LookUpID)
        {
            return Portal.Fetch(LookUpID);
        }

I changed it to the following. 

        public async Task<LookUpTableInfo> GetkLookUpInfoAsync(Guid LookUpID)
        {
            return await Portal.FetchAsync(LookUpID);
        }

I will run some test and get back to you. 
Thank you for your help.

**Comment by jhuerta3** [comment, marked as answer]
After doing some research I found the issue it was a combination of two things. First, I had to correct my code to the following which still needs some work:

        protected override async void OnPropertyChanged(IPropertyInfo propertyInfo)
        {
            base.OnPropertyChanged(propertyInfo);
            if (propertyInfo == StatusIDProperty)
            {
                if (StatusID != Guid.Parse("00000000-0000-0000-0000-000000000000"))
                {
                    LookUpTableInfoPortal = ApplicationContext.GetRequiredService<IDataPortal<LookUpTableInfo>>();
                    LookUpTableFactory objLookFactory = new LookUpTableFactory(LookUpTableInfoPortal);
                    Task<LookUpTableInfo> objStatusTask = objLookFactory.GetkLookUpInfoAsync(StatusID);
                    LookUpTableInfo objStatus = await objStatusTask;

                    ActiveFlag = !objStatus.LookUpName.ToLower().Equals("Obsolete".ToLower()) &&
                                    !objStatus.LookUpName.ToLower().Equals("Damaged".ToLower());
                }
            }
        }

Then I realized that when I debug on a break point my IDE would freeze. I thought the code wasn’t working. After I removed all the break points it ran without any issues. I got another problem but I will ask another question, so it has its own topic.

**Reply by rockfordlhotka** [reply]
I've had my share of issues with debugging in Blazor projects too. No good solution to offer, just my empathy.
